### PR TITLE
executor: split hashjoin part4

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/39061

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/join.go [hunk 1]
```diff
@@ -69,6 +78,10 @@ type probeSideTupleFetcher struct {
 }
 
 type probeWorker struct {
+	*hashJoinCtx
+	sessCtx sessionctx.Context
```
Comment (wshwsh12) [inline executor/join.go:82]: Move to hashJoinCtx?

File: executor/join.go [hunk 2]
```diff
@@ -81,42 +94,32 @@ type probeWorker struct {
...
-	probeWorkers      []probeWorker
+	probeWorkers          []probeWorker
+
 	buildSideExec     Executor
 	buildSideEstCount float64
-	outerFilter       expression.CNFExprs
 	probeKeys         []*expression.Column
 	probeNAKeys       []*expression.Column
 	buildKeys         []*expression.Column
-	buildNAKeys       []*expression.Column
```
Comment (wshwsh12) [inline executor/join.go:99]: Why only move `buildNAKeys` into hashJoinCtx?

File: executor/join.go [hunk 3]
```diff
@@ -443,84 +447,84 @@ func (e *HashJoinExec) waitJoinWorkersAndCloseResultChan() {
...
 			break
 		}
 		start := time.Now()
-		if e.useOuterToBuild {
-			ok, joinResult = e.join2ChunkForOuterHashJoin(workerID, probeSideResult, hCtx, e.probeWorkers[workerID].rowContainerForProbe, joinResult)
+		if w.useOuterToBuild {
+			ok, joinResult = w.join2ChunkForOuterHashJoin(probeSideResult, hCtx, joinResult)
 		} else {
-			ok, joinResult = e.join2Chunk(workerID, probeSideResult, hCtx, e.probeWorkers[workerID].rowContainerForProbe, joinResult, selected)
+			ok, joinResult = w.join2Chunk(probeSideResult, hCtx, w.rowContainerForProbe, joinResult, selected)
```
Comment (wshwsh12) [inline executor/join.go:504]: Why `join2ChunkForOuterHashJoin` remove the arg `e.probeWorkers[workerID].rowContainerForProbe` but `join2Chunk` not?

File: executor/join.go (hunk 2 above)
Comment (AilinKid) [reply executor/join.go:99]: I think the only usage below https://github.com/pingcap/tidb/pull/39155/files#diff-ac6f3fc5c91a1d77229d11140f9c8a8b561eb33ce595fc2d0ffd5d912cf8e3abR893 can use a bool to substitute

File: executor/join.go (hunk 1 above)
Comment (XuHuaiyu) [reply executor/join.go:82]: I'll move it if needed in the following PRs.

**Review by wshwsh12 — APPROVED** [review]

**Pull request merged**